lib/jellyfish/notifier.ex

```diff
@@ -60,34 +57,85 @@ defmodule Jellyfish.Notifier do
     connect(:start, opts)
   end
 
+  @typedoc """
+  Option values for `subscribe/2`.
+
+  * `:process` - pid of the process that will receive notifications, `self()` by default.
+  * `:room_id` - id of the room. By default notifications about all of the rooms are received,
+  if this parameter is passed, notifications will be filtered only to this specific room.
+  """
+  @type subscribe_options :: [process: Process.dest(), room_id: Room.id()]
+
+  @doc """
+  Subscribes the process to receive server notifications.
+
+  Notifications are sent to the process in a form of `{:jellyfish, msg}`,
+  where `msg` is one of the structs defined in `lib/proto/jellyfish/server_notifications.pb.ex`,
+  for example `{:jellyfish, %Jellyfish.ServerMessage.RoomCrashed{room_id: "some_id"}}`.
+
+  For information about options, see `t:subscribe_options/0`.
+  """
+  @spec subscribe(GenServer.server(), subscribe_options()) :: :ok
+  def subscribe(notifier, opts \\ []) do
+    process = Keyword.get(opts, :process, self())
+    room = Keyword.get(opts, :room_id, :all)
+
+    WebSockex.cast(notifier, {:subscribe, process, room})
+  end
+
+  @doc """
+  Unsubscribe from receiving notifications requested by calling `subscribe/2`.
+
+  Stops the `Notifier` from sending any notifications to `process`.
+  """
+  @spec unsubscribe(GenServer.server(), Process.dest()) :: :ok
+  def unsubscribe(notifier, process \\ self()) do
+    WebSockex.cast(notifier, {:unsubscribe, process})
+  end
+
   @impl true
   def handle_frame({:binary, msg}, state) do
     %ServerMessage{content: {_type, notification}} = ServerMessage.decode(msg)
+    handle_notification(notification, state)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_cast({:subscribe, pid, room_id}, state) do
+    state =
+      state.subscriptions
+      |> Map.get(room_id, MapSet.new())
+      |> MapSet.put(pid)
+      |> then(&put_in(state, [:subscriptions, room_id], &1))
```

Comment:
I'd replace that pipe with `update_in`:
```suggestion
      update_in(state.subscriptions[room_id], fn
        nil -> MapSet.new()
        set -> MapSet.put(set, pid)
      end)
```